src/rwe/GameScene.cpp

```diff
@@ -1980,6 +1984,23 @@ namespace rwe
         // run resource updates once per second
         if (simulation.gameTime % GameTime(SimTicksPerSecond) == GameTime(0))
         {
+            // recalculate max energy and metal storage
+            for (auto& player : simulation.players)
+            {
+                player.maxEnergy = Energy(0);
+                player.maxMetal = Metal(0);
+            }
+
+            for (auto& entry : simulation.units)
+            {
+                auto& unit = entry.second;
+                if (!unit.isBeingBuilt())
+                {
+                    getPlayer(unit.owner).maxMetal += unit.metalStorage;
+                    getPlayer(unit.owner).maxEnergy += unit.energyStorage;
+                }
+            }
+
```

Comment:
I'm not convinced this is the right way to do it. Since total energy and metal storage will only change whenever a unit is spawned, is transferred to different owner, or dies, I would think it better to update the player's totals whenever one of these events happens, rather than recomputing it every second like this. Was there something that motivated you to go with this design? (e.g. are you convinced this is how TA does it?)

Reply:
I don't know how TA handles this. I thought about the approach you suggest as well. The main reason I went with this instead was that I thought it made the code slightly simpler. All of the code is in one place, and it's fairly easy to quickly get an overview over it, and thus verify that it's correct (it's maybe a bit more unlikely that a future refactoring regarding say e.g. how units are destroyed accidentally break the storage calculations). Which might be a bit harder if this logic has to be spread out at several places (spawning, transferring, and dying as you say).

But I'm happy to try to change it to what you suggest! Are your main concerns about performance, or more about what conceptually makes sense?

Btw, RWE is a really cool project and I'm amazed over what you've been able to achieve! 😃

Reply:
I think performance is a decent argument in favour of doing the updates on spawn/death since that's an awful lot of redundant work to be doing every second. If TA only updates this every second then that would be a good argument for keeping things the way they are, but I haven't tested this behaviour.

Reply:
It's possible that both things are happening, in a sense: a completed storage unit queues an update to the storage values somewhere, which then gets processed by the game once per second. Would that be an appropriate compromise?

Reply:
Thanks both of  you for your input! I did some quick testing (manually spawning metal storages with `+armmstor`, setting the speed to -9 and pausing/unpausing) and it looks like the storage values (at least what's shown in the top of the UI) updates once per second (when the `+clock` ticks a second) and not immediately when the unit spawns. So then the approach @Skirmisher suggest might address the performance concerns you have @MHeasell while also keeping the behaviour as similar to TA as possible? Or would it be better to immediately update the values the same frame the unit is spawned, dies etc?

Reply:
Nah I think the implementation you've already written is fine then.

Just squash your fixup commit and I'll happily merge this.

Reply:
Alright, done! 👍

Reply:
⭐